src/main/java/io/netty/incubator/codec/http3/Http3ServerPushStreamManager.java

```diff
@@ -0,0 +1,289 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.incubator.codec.http3;
+
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandler;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.channel.socket.ChannelInputShutdownReadComplete;
+import io.netty.incubator.codec.quic.QuicChannel;
+import io.netty.incubator.codec.quic.QuicStreamChannel;
+import io.netty.incubator.codec.quic.QuicStreamChannelBootstrap;
+import io.netty.incubator.codec.quic.QuicStreamType;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.UnaryOperator;
+
+import static io.netty.incubator.codec.http3.Http3.maxPushIdReceived;
+import static io.netty.incubator.codec.http3.Http3CodecUtils.connectionError;
+import static io.netty.incubator.codec.http3.Http3ErrorCode.H3_ID_ERROR;
+import static io.netty.util.internal.PlatformDependent.newConcurrentHashMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
+
+/**
+ * A manager for <a href="https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-push-streams">push streams</a>
+ * for a server. New push streams can be initiated using the various {@code newPushStream} methods. It is required to
+ * add the {@link ChannelHandler} returned from {@link #controlStreamListener()} to the {@link QuicChannel} associated
+ * with this manager.
+ */
+public final class Http3ServerPushStreamManager {
+    private static final AtomicLongFieldUpdater<Http3ServerPushStreamManager> nextIdUpdater =
+            newUpdater(Http3ServerPushStreamManager.class, "nextId");
+    private static final Object CANCELLED_STREAM = new Object();
+    private static final Object PUSH_ID_GENERATED = new Object();
+    private static final Object AWAITING_STREAM_ESTABLISHMENT = new Object();
+
+    private final QuicChannel channel;
+    private final ConcurrentMap<Long, Object> pushStreams;
+    private final ChannelInboundHandler controlStreamListener;
+
+    private volatile long nextId;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param channel for which this manager is created.
+     */
+    public Http3ServerPushStreamManager(QuicChannel channel) {
+        this(channel, 8);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param channel for which this manager is created.
+     * @param initialPushStreamsCountHint a hint for the number of push streams that may be created.
+     */
+    public Http3ServerPushStreamManager(QuicChannel channel, int initialPushStreamsCountHint) {
+        this.channel = requireNonNull(channel);
```

Comment:
I recommend specifying the parameter name when calling `requiredNonNull` for a parameter, so a user/dev knows quickly what was `null`.
```suggestion
        this.channel = requireNonNull(channel, "channel");
```